.circleci/cimodel/pytorch_build_definitions.py

```diff
@@ -0,0 +1,369 @@
+                 dependent_tests=None,
+                 parent_build=None):
+
+        self.distro = distro
+        self.pyver = pyver
+        self.parms = parms
+        self.cuda_version = cuda_version
+        self.is_xla = is_xla
+        self.restrict_phases = restrict_phases
+        self.gpu_resource = gpu_resource
+        self.dependent_tests = dependent_tests or []
+        self.parent_build = parent_build
+
+    def get_parms(self):
+        leading = ["pytorch"]
+        if self.is_xla:
+            leading.append(DockerHide("xla"))
+
+        cuda_parms = []
+        if self.cuda_version:
+            cuda_parms.extend(["cuda" + self.cuda_version, "cudnn7"])
+        return leading + ["linux", self.distro] + cuda_parms + self.parms
+
+    # TODO: Eliminate this special casing in docker paths
+    def gen_docker_image_path(self):
+
+        build_env_pieces = list(map(str, filter(lambda x: type(x) is not DockerHide, self.get_parms())))
+        base_build_env_name = "-".join(build_env_pieces)
+
+        return miniutils.quote(DOCKER_IMAGE_PATH_BASE + base_build_env_name + ":" + str(DOCKER_IMAGE_VERSION))
+
+    def get_build_job_name_pieces(self, build_or_test):
+        return self.get_parms() + [build_or_test]
+
+    def gen_build_name(self, build_or_test):
+        return ("_".join(map(str, self.get_build_job_name_pieces(build_or_test)))).replace(".", "_")
+
+    def get_dependents(self):
+        return self.dependent_tests
+
+    def gen_yaml_tree(self, build_or_test):
+
+        build_job_name_pieces = self.get_build_job_name_pieces(build_or_test)
+
+        build_env_name = "-".join(map(str, build_job_name_pieces))
+
+        env_dict = OrderedDict([
+            ("BUILD_ENVIRONMENT", build_env_name),
+            ("DOCKER_IMAGE", self.gen_docker_image_path()),
+        ])
+
+        if self.pyver:
+            env_dict["PYTHON_VERSION"] = miniutils.quote(self.pyver)
+
+        if build_or_test == "test" and self.cuda_version:
+            env_dict["USE_CUDA_DOCKER_RUNTIME"] = miniutils.quote("1")
+
+        d = {
+            "environment": env_dict,
+            "<<": "*" + "_".join(["pytorch", "linux", build_or_test, "defaults"]),
+        }
+
+        if build_or_test == "test":
+            resource_class = "large"
+            if self.gpu_resource:
+                resource_class = "gpu." + self.gpu_resource
+
+                if self.gpu_resource == "large":
+                    env_dict["MULTI_GPU"] = miniutils.quote("1")
+
+            d["resource_class"] = resource_class
+
+        return d
+
+    def gen_workflow_yaml_item(self, phase):
+
+        if self.is_xla or phase == "test":
+            val = OrderedDict()
+            if self.is_xla:
+                val["filters"] = {"branches": {"only": ["master"]}}
+
+            if phase == "test":
+                dependency_build = self.parent_build or self
+                val["requires"] = [dependency_build.gen_build_name("build")]
```

Comment:
When merging the caffe2 and pytorch jobs, it might be convenient for a while to make a caffe2 test job dependent on a pytorch build job. This way we could quickly dedup the repeated build of pytorch in the caffe2 build job, and just run the caffe2 tests off of a completed pytorch build jobb